libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/MapboxTripStarter.kt

```diff
@@ -0,0 +1,196 @@
+ *
+ * In order to share the state between an App and Android Auto, the instance of this class should
+ * be shared. That will be done automatically if you use [getRegisteredInstance].
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class MapboxTripStarter internal constructor() : MapboxNavigationObserver {
+
+    private val optionsFlow = MutableStateFlow(MapboxTripStarterOptions.Builder().build())
+    private val stateFlow = MutableStateFlow(MapboxTripStarterState())
+    private var replayRouteTripSession: ReplayRouteSession? = null
+    private var mapboxNavigation: MapboxNavigation? = null
+
+    @VisibleForTesting
+    internal var coroutineScopeProvider: () -> CoroutineScope = {
+        CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
+    }
+    private lateinit var coroutineScope: CoroutineScope
+
+    /**
+     * Signals that the [mapboxNavigation] instance is ready for use.
+     * @param mapboxNavigation
+     */
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        this.mapboxNavigation = mapboxNavigation
+        coroutineScope = coroutineScopeProvider()
+
+        // Initialize the options to be aware of the location permissions
+        val context = mapboxNavigation.navigationOptions.applicationContext
+        val granted = PermissionsManager.areLocationPermissionsGranted(context)
+        stateFlow.update { it.copy(isLocationPermissionGranted = granted) }
+
+        // Observe any changes to the options
+        coroutineScope.launch {
+            combine(optionsFlow, stateFlow, ::Pair).collect {
+                onStarterOptionsChanged(mapboxNavigation, it.first, it.second)
+            }
+        }
+    }
+
+    /**
+     * Signals that the [mapboxNavigation] instance is being detached.
+     * @param mapboxNavigation
+     */
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        coroutineScope.cancel()
+        onTripDisabled(mapboxNavigation)
+        this.mapboxNavigation = null
+    }
+
+    /**
+     * Get the options that are currently set.
+     */
+    fun getOptions(): MapboxTripStarterOptions = optionsFlow.value
+
+    /**
+     * Set new options.
+     */
+    fun setOptions(options: MapboxTripStarterOptions) = apply {
+        this.optionsFlow.value = options
+    }
+
+    /**
+     * Get the location permission state. This may not reflect the actual permissions. Used for
+     * verifying the tests and will likely be removed.
+     */
+    @VisibleForTesting
+    internal fun getLocationPermissionGranted(): Boolean =
+        stateFlow.value.isLocationPermissionGranted
+
+    /**
+     * Set the location permission state. [MAPBOX_TRIP_STARTER_FOLLOW_DEVICE] will not work unless
+     * location permissions have been granted. If the location permissions are not granted and
+     * this is set to true, an error will be thrown.
+     */
+    @Throws(IllegalStateException::class)
+    fun setLocationPermissionGranted(granted: Boolean) {
```

Comment:
Why do we expose a way of setting inconsistent state? What if we rename this method into `onLocationPermissionsStateChanged()` without any arguments and for the actual boolean value we just check the `PermissionsManager`? We do it either way, why bother the user with checking it as well?

Reply:
Permissions are accepted through an Activity. PermissionManager requires a check, rather than an observer. So for now, we need Activity.onLocationPermissionGranted->MapboxTripProgress

I'd prefer PermissionManager to do that, but it does not yet. So I'm considering this current approach temporary

Reply:
The comments on the class say "The [MapboxTripStarter] is not able to observe when location
 * permissions change, so you must notify it of changes"

Reply:
I understand that.
I mean let's change this method into sth like:
```
fun onLocationPermissionsStateChanged() {
    val granted = PermissionsManager.areLocationPermissionsGranted(context)
    stateFlow.value = stateFlow.value.copy(isLocationPermissionGranted = granted)
}
```
And the user will invoke `onLocationPermissionsStateChanged()` whenever they get permissions result (regardless of the fact if it's granted or not, we check the boolean flag).

Reply:
Yeah thanks I agree. Adding this check and exception throwing, is only making things complicated. Am replacing it with a `refreshLocationPermissions()` function.